### ISSUE TITLE
Do not create string objects from consumerTag, exchange and routingKey, or get them from a string cache.

### ISSUE THREAD

**Issue by zgabi** [rabbitmq/rabbitmq-dotnet-client#1231]
In our real application the most allocated objects are Strings. Most of them are allocated in RabbitMQ Client:

![image](https://user-images.githubusercontent.com/3104253/175136450-ad2111d7-8cbe-44b3-b596-f0bc8d3edb0f.png)

They usually contains the same value.

Please consider using `ReadOnlySpan<char>` or `ReadOnlySpan<byte>` with a pooled array in the background similar to the data.
Alternatively get the strings from a cache table (maybe separate table for each "string type" (consumerTag, exchange...)), since most of them are the same.

By the way in the screenshot the 3rd, 4th, 5th and 6th allocations also belongs to RabbitMQ Client, but maybe some of them are already solved in the main branch. (I'm using the latest 6.x package)
The 2nd one is WFP, I have a PR for that.

**Comment by zgabi** [comment]
I don't undertand your question. We have 1 consumer in this process.

**Comment by zgabi** [comment]
Which one do you prefer?

1. Receive ReadOnlySpan<char> in the `IBasicConsumer`'s Handle* methods (e.g `HandleBasicDeliver`), so it is already UTF-8 decoded characters 
2. Receive ReadOnlySpan<char> in the `IBasicConsumer`'s Handle* methods (e.g `HandleBasicDeliver`), the raw bytes
3. Do not change the interface, but cache the strings in WireFormatting.ReadShortstr.. maybe this one is slower, because the cached strings should be compared with the received data. But do not require to change the interface.

**Comment by michaelklishin** [comment]
@zgabi my point is that caching is always harder than it sounds. What would be the gains (in terms of memory footprint and allocations) to a realistic consumer (application) like yours, or even a basic tutorial-style one.

Previously all efficiency improvements in this client were driven by profiler results. Sure,
you already have some but I am curious about the difference it makes.

**Comment by zgabi** [comment]
> I'd like to get feedback from the "usual gang", aka @bollhals, @bording, @danielmarbach and @stebet if they have time. I'm a bit surprised nobody has reported this issue in their use of this library.

Because they are just some small allocations. 2-3 small strings / message ("exchange" is empty string for me, which is not allocated)
And everybody is happy since a lot of other allocations were removed in the past 3-4 years. (Me, too) Version 6.x is much better and faster than 5.x was 4 years ago. And I saw that the main branch has more new related improvelemts (BasicProperties is a struct, etc...) 

We are sending a lot of messages (~200/ sec), because it is a requirement for us to send everything between the server and the clients through RabbitMQ... even video streams.
This string allocation problem is also not a big issue for us... 1.8 M messages allocations / 38 minutes. GC can handle it :)

I reporteed this issue, since it is the top 1 in the memory profiler.

**Comment by zgabi** [comment]
Yes, caching is harder than simply passing spans to the user.

For exmaple:
- caculating some hash from the data
- put it to a thread static dictionary (with this the string will exist only 1 time / thread, which is quite good, and does not need locking).

So probably this is slower than just simply creating a new string.

So I'd prefer passing span (or memory) of byte to the user.... similar to the data.
It needs the modification of the `IBasicConsumer`, but does not have any other disadvantages. It will be even faster, since no UTF-8 decoding is needed. And no string allocation, although it is fast...

**Comment by michaelklishin** [comment]
@zgabi we can do this for 7.0. Option 2 above is more in the spirit of RabbitMQ and clients but I can see how many might vote for option 1.

**Comment by bollhals** [comment]
> In our real application the most allocated objects are Strings. Most of them are allocated in RabbitMQ Client:

Yeah the strings always were a bit of a problem, but we've focused on everything else so far. 

> Which one do you prefer?
> 
> 1. Receive `ReadOnlySpan<char>` in the `IBasicConsumer`'s Handle* methods (e.g `HandleBasicDeliver`), so it is already UTF-8 decoded characters
> 2. Receive `ReadOnlySpan<byte>` in the `IBasicConsumer`'s Handle* methods (e.g `HandleBasicDeliver`), the raw bytes
> 3. Do not change the interface, but cache the strings in WireFormatting.ReadShortstr.. maybe this one is slower, because the cached strings should be compared with the received data. But do not require to change the interface.

Regarding the options
ROS<> in general is problematic for as long as we want to support async methods. 
We can think about ROM<>.

Caching, I've actually 1.5 years ago started experimenting with option 3 back then (see commit [here](https://github.com/bollhals/rabbitmq-dotnet-client/commit/8bbac10dbd841bee0dfd559abf7362b8df8a2cc0)) and even saw gains in overall performance if the string I think was > 16 chars. But it's been a while, so one would had to check again.

In principal the API is a complex topic and depends which usecase we want to support.

- Caching is the default easy to use for the customer API that the user already knows. (Though fine tuning might require some new API, e.g. for cleaning up the cache)
- ROM<char> is still kind of convenient to use, but often you have to have a own copy / ToArray() it, where you'd loose all gains again. (E.g. all ConsumeTag are stored in the consumer, so we'd have to adapt the API there as well and allocate here to store the information.)
- ROM<byte> on to of the above, we'd save on the encoding. But pay the price of usability. In some cases (especially where you don't use it) you have a lot of gain, but on others where you do want to access  it en encode, it looks like it's more a step in the wrong  direction

=> So for me the caching option looks like the best option if the API can somewhat be simple enough for the common use case.

**Comment by zgabi** [comment]
What about distinguish the consumerTag from the another 2 strings. As you wrote consumerTags are problematic, they are stored in dictionary and hashset. Multiple handlers receve it.
It is quite easy to get rid of the another two to change it to `ROM<byte>`  or similar custom struct which has an overridden `ToString`... so the user's can easily get the string value.

I've investigated a little bit.

This is measure from the current main branch:
![image](https://user-images.githubusercontent.com/3104253/175482106-04d621f5-35c1-4df5-bfd5-af2ec37602ae.png)

(Btw: you can see that everything other disappeared from the list which was in v 6.4 in my first comment)

This measure is when I change the exchange and routingKey to `ROM<byte>`:
![image](https://user-images.githubusercontent.com/3104253/175486026-3770d29a-5776-4d0a-b9b8-e3728e8dc182.png)

Changes: https://github.com/zgabi/rabbitmq-dotnet-client/commit/a6ac5c47e3407b53dcee1a4645aa83df07cc07be

I tried to make similar time periods for the measure, but check the ratio between the number of the objects. In the latest main code there are almost the same number ofd strings as WPF EventKeys. In the modified code about half of them.

I could change the ROM to another struct with an overridden ToString. It won't affect the memory allocations, and also not really changes the performance. So the user can get the string... but he has the opportunity to convert to `Span<char>` to without allocation if he also cares it.

**Comment by danielmarbach** [comment]
I haven't looked in depth into the code base for a while... I was wondering if we have already looked at places where we can benefit from the low allocation APIs like `string.Create` to give use some gains there before we go into the complexity of string caching?

**Comment by zgabi** [comment]
@danielmarbach `string.Create` is also creating a new string object :) This issue is not about string concetanation where `string.Create` would be useful. There are no concatenation in this case.

**Comment by danielmarbach** [comment]
yeah I'm aware of that. My point was not about the specific problem but more about making sure where we use string and concatenation we do it efficiently before we go into complex string caching.

**Comment by zgabi** [comment]
As far as I saw there are no string concatenation issues in the code. After the connection is established 100% of the new strings in RabbitMQ client are cunsumerTags, excanges and routingKeys.

The only string concatientaion in this library is in the ToString method of the ShutdownEventArgs.. which is (according to the method comment's) only for debugging:
https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/6b1a06bd429f395891a3230cad92e674dcbbb0d2/projects/RabbitMQ.Client/client/api/ShutdownEventArgs.cs#L97-L108

**Comment by zgabi** [comment]
Created a test commit for removing the string allocation.
Since the consumers are stored in a dictionary, and the dictionary lookup is needed anyway, RabbitMQ client can store the the known clientTag strings in that dictionary. I haven't measured, but probably this one is slightly faster, because the "utf8 data" is about half than the "string data", so the hash calcualtion and comparsion should be faster.
The string allocation is needed only when there is no consumer for the consumerTag.
For me it removes 100% of the string allocations. I don't know whether it is a problem for anybody that the allocations are not removed when consumer does not exists. (How can that happen?)

Commit:
 https://github.com/zgabi/rabbitmq-dotnet-client/commit/4b67b238b627abef901e448099e4672bd8c498fd

Memory profiler result after both commits:
![image](https://user-images.githubusercontent.com/3104253/175613529-dc9326a2-9cb1-4427-922b-40c3b7c63d3b.png)

**Comment by danielmarbach** [comment]
FYI Ben Adams wrote once an intern pool https://github.com/benaadams/Ben.StringIntern.

Yet it seems by the first look your approach is elegant and low effort. Need to look closer though a bit next week

**Comment by bollhals** [comment]
One thing you need to take care of if you store it in a dictionary or cache it, is that the backing storage is a rented array, hence it will be returned and potentially overwritten, which changes the content of your ROM<byte>

I'll try to look into the options and provide a bit more context.

Btw, we already have this [CachedString](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/client/api/CachedString.cs) class here, the intend when introducing it was cases like this.

**Comment by zgabi** [comment]
Yes, I create a new byte[] in the AddConsumer method. No rented array is used after it is returned to the pool.

**Comment by bollhals** [comment]
If I'm not mistaken then you return the array right after you call dispatch. Whereas for de body array we pass it on to the dispatcher.

**Comment by zgabi** [comment]
Oh, so you mean the excahnge and routingKey.

Yes.

I've created new commits, fixed some other things which was commented by @danielmarbach (See comments here: https://github.com/zgabi/rabbitmq-dotnet-client/commit/4b67b238b627abef901e448099e4672bd8c498fd)

New commits:
https://github.com/zgabi/rabbitmq-dotnet-client/commit/926b2f848576e9afc7033ca3d5e3728538458ce6
and 
https://github.com/zgabi/rabbitmq-dotnet-client/commit/fc636a2c8a6dfc7419dc3b7a06cd29f851a8f755

**Comment by bording** [comment]
> I've created new commits, fixed some other things which was commented by @danielmarbach (See comments here:

You should make your commits on a separate branch and open a draft PR. It makes it much easier to see what you're doing that way.

**Comment by zgabi** [comment]
Ok, created a draft PR (#1233)

**Comment by zgabi** [comment]
Yes, all the methods which has routingKey / exchange should accept `ROM<byte>`,too.. even more, that should be the prefered use. 
For compatibility they can have an overload which accepts string, but internally they should only convert the received parameter to byte[]. So the users should know that when they use the string parameter version, it is slower.
This makes the CachedString and some relates classes unnecessary. (For example the class called BasicPublis... always the BasicPublishMemory will be used - which should be renamed to BasicPublish)

Instead of accepting (and providing in the Handle* methods) `ROM`, the parameter type could be a "CustomString" (todo: give a good name) struct, which internally similar to the `ROM<byte>`, but it will be obvoius to the users that they are strings. Can for example explicit/implicit cast from/to string or call ToString().

In this case it will be enough to have only 1 parameter combinataion for for example the BasicPublish method

`        void BasicPublish<TProperties>(CustomString exchange, CustomString  routingKey, ref TProperties basicProperties, ReadOnlyMemory<byte> body = default, bool mandatory = false)
`

And the users can call it with strings (implicit cast):
`        BasicPublish("myexchange", "myRoutingKey", ref basicProperties, body)
`

And the users who are interested in the performance will "cache" the CustomString stuct.

**Comment by michaelklishin** [comment]
@zgabi in the protocol, the type used by those fields is called "short strings". It can be a decent type name to use in the context of this client.

**Comment by bollhals** [comment]
The default up until now was and still is the string case, so we should "try" not to make this case worse, but rather provide a even better alternative. (The CachedString overload was just recently added and can probably easily be modified if needed)

Nesting structs in structs will have it's own challenges and inefficiencies, but if wanted it can be managed. But I'd hold off implicit conversions to this type in general, and from this type if they allocate. (E.g. No implicit converstion to string, if it means allocating a string. then I rather have an explicit ConvertToString or similar)

**Comment by zgabi** [comment]
I created a new commit which is using CachedStrings for every exhange and routingKey (sometimes the parameter name is "name" or "source", but it contains exchange or routingKey)

This branch and commit does not contain the counsumerTag change as requested:
https://github.com/zgabi/rabbitmq-dotnet-client/commit/2e71e8b5c7f23514b62aef17099593232379c84f

All public method has both string and CachedString overloads. Internal is only CachedString (when it is enough)

Should I create a new PR?

**Comment by zgabi** [comment]
Any news on this issue?

**Comment by michaelklishin** [comment]
FTR, the PR was submitted as https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1233.

**Comment by lukebakken** [comment]
@zgabi as you can see, in #1233 I have brought your code in-line with `main`.

Could you please provide _precise_ instructions, including code if necessary, for me to see the benefit of this PR? Assume I have never done memory profiling (because it has been so long I don't remember the details). Thank you.

**Comment by lukebakken** [comment]
@zgabi @bollhals @michaelklishin 

Hi everyone, now that I've done some work on #1233 and have gotten more familiar with the rationale behind this discussion and associate PR, I have the following -

* I'd really like to get version 7 of this library done May of 2024. Do the benefits to this change justify extending the release date?
* Currently, #1233 modifies the public API for handling deliveries such that exchange and routing key strings are `ReadOnlyMemory<byte>`. I agree with @bollhals that the public API _must_ be consistent in this library, so I am leaning towards re-naming `CachedString` to `ShortStr` (as is named in the AMQP 091 spec) and using it consistently through the API for all `string` arguments. Thoughts?
